book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -165,49 +220,232 @@ each rooted at the highest finalized block. Each chain consists of a map from
+    - Add hash to `height_by_hash`
+    - for each `transaction` in `block`
+      - add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
+    - Add new utxos and remove consumed utxos from `self.utxos`
+    - Add anchors to the appropriate `self.<version>_anchors`
+    - Add nullifiers to the appropriate `self.<version>_nullifiers`
+    - Add work to `self.partial_cumulative_work`
+
+#### `pub fn pop_root(&mut self) -> Arc<Block>`
+
+Remove the lowest height block of the non-finalized portion of a chain.
+
+1. Remove the lowest height block from `self.blocks`
+
+1. Update cummulative data members
+    - Remove the block's hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Remove new utxos from `self.utxos`
+    - Remove the anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+
+1. Return the block
+
+**Note**: We do not subtract work from `self.partial_cummulative_work`. This
+is to make make the ordering of chains stable while finalizing blocks.
+
+#### `pub fn fork(&self, new_tip: block::Hash) -> Option<Self>`
+
+Fork a chain at the block with the given hash, if it is part of this chain.
+
+1. If `self` does not contain `new_tip` return `None`
+
+2. Clone self as `forked`
+
+3. While the tip of `forked` is not equal to `new_tip`
+   - call `forked.pop_tip()` and discard the old tip
+
+4. Return `forked`
+
+#### `fn pop_tip(&mut self) -> Arc<Block>`
+
+Remove the highest height block of the non-finalized portion of a chain.
+
+1. Remove the highest height `block` from `self.blocks`
+
+1. Update cummulative data members
+    - Remove the corresponding hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Add consumed utxos and remove new utxos from `self.utxos`
+    - Remove anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+    - Subtract work from `self.partial_cumulative_work`
+
+1. Return the block
 
-Otherwise, attempt to perform contextual validation checks and the commit the
-given block to the state. The exact list of contextual validation checks will
-be specified in a later RFC. If contextual validation checks succeed, commit
-the new blocks to the non-finalized state as described below. Next, if the
-resulting non-finalized chain is longer than 100 blocks, the oldest block is
-now past the reorg limit. Remove it from the non-finalized state and commit
-it as a finalized block, as described in the next section.
+#### `Ord`
 
-XXX: fill in details of non-finalized state.
+The `Chain` type also implements `Ord` for reorganizing chains. First chains
+are compared by their `partial_cummulative_work`. Ties are then broken by
+comparing `BlockHeaderHashes` of the tips of each chain.
 
```

Comment:
Let's explain why we made the decision about the block hash tie-breaker:

```suggestion

**Note**: Unlike `zcashd`, Zebra does not use block arrival times as a tie-breaker for the best tip. Since Zebra downloads blocks in parallel, download times are not guaranteed to be unique. Using the `block::Hash` provides a consistent tip order. (As a side-effect, the tip order is also consistent after a node restart, and between nodes.)
```